File: general-interview-prep/javascript/backspaceCompare.js
/**
 * @param {string} S
 * @param {string} T
 * @return {boolean}
 */
var backspaceCompare = function(S, T) {
    let first = [];
    let second = [];

    for (let char of S) {
        if (char !== "#") {
            first.push(char);
        } else if (first.length !== 0) {
            first.pop();
        }
    }

    for (let char of T) {
        if (char !== "#") {
            second.push(char);
        } else if (second.length !== 0) {
            second.pop();
        }
    }

    while(first.length !== 0) {
        let current = first.pop();
        if (second.length === 0 || second.pop() !== current) {
            return false;
        }
    }

    return true;
};

File: general-interview-prep/golang/backspaceCompare.go
package main

func backspaceCompare(S string, T string) bool {
	s, t := toArray(S), toArray(T)
	if len(s) != len(t) {
		return false
	}
	for i := 0; i < len(s); i++ {
		if s[i] != t[i] {
			return false
		}
	}
	return true
}

func toArray(S string) []byte {
	var s []byte
	for i := 0; i < len(S); i++ {
		switch S[i] {
		case '#':
			if len(s) > 0 {
				s = s[:len(s)-1]
			}
		default:
			s = append(s, S[i])
		}
	}
	return s
}

func main() {
  fmt.Println(backspaceCompare("ab#c", "ad#c"))
}
